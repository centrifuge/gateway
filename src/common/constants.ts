const API_BASE = '/api';

const baseUsersRoute = `${API_BASE}/users`;

const userRoutes = {
  base: baseUsersRoute,
  login: `${baseUsersRoute}/login`,
  logout: `${baseUsersRoute}/logout`,
  invite: `${baseUsersRoute}/invite`,
  register: `${baseUsersRoute}/register`,
};

export const ROUTES = Object.freeze({
  API_BASE,
  INVOICES: `${API_BASE}/invoices`,
  USERS: userRoutes,
  CONTACTS: `${API_BASE}/contacts`,
  WEBHOOKS: `${API_BASE}/webhooks`,
  PURCHASE_ORDERS: `${API_BASE}/purchase_orders`,
});

export enum PERMISSIONS {
  CAN_FUND_INVOICES = 'can_fund_invoices',
  CAN_CREATE_INVOICES = 'can_create_invoices',
  CAN_MANAGE_USERS = 'can_manage_users',
  CAN_MANAGE_ACCOUNTS = 'can_manage_accounts',
}
