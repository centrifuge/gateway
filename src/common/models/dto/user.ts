export class User {
  constructor(
    readonly username: string,
    readonly password: string,
    readonly id: string,
  ) {}
}
