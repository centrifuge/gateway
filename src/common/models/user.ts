import { PERMISSIONS } from '../constants';
import { Document, DOCUMENT_ACCESS } from './document';

export interface IUser {
  name: string;
  password?: string;
  email: string;
  _id?: string;
  account: string;
  permissions: PERMISSIONS[];
  schemas?: string[];
  enabled: boolean;
  invited: boolean;
}

export class User implements IUser {
  name: string = '';
  password?: string = '';
  email: string = '';
  _id?: string;
  account: string = '';
  permissions: PERMISSIONS[] = [];
  schemas: string[] = [];
  enabled: boolean;
  invited: boolean;
}

export const canWriteToDoc = (user: User | null, doc?: Document): boolean => {
  if(!user || !doc) return false;
  return accountHasDocAccess(user.account, DOCUMENT_ACCESS.WRITE, doc);
};

export const canReadDoc = (user: User | null, doc?: Document): boolean => {
  if(!user || !doc) return false;
  return accountHasDocAccess(user.account, DOCUMENT_ACCESS.READ, doc);
};

export const accountHasDocAccess = (account: string, access: DOCUMENT_ACCESS, doc?: Document): boolean => {
  return !!(
    doc &&
    doc.header &&
    doc.header[access] &&
    Array.isArray(doc.header[access]) &&
    doc.header[access]!.find(
      ac => ac.toLowerCase() === account.toLowerCase(),
    )
  );
};


export const canCreateDocuments = (user: User): boolean => {
  return (
    user.permissions.includes(PERMISSIONS.CAN_MANAGE_DOCUMENTS)
    && user.schemas.length > 0);
};


export const canSignFunding = (user:User | null, doc?:Document):boolean => {
  if(!user) return false;
  return !!(
    doc &&
    doc.attributes &&
    doc.attributes.funding_agreement &&
    Array.isArray(doc.attributes.funding_agreement) &&
    doc.attributes.funding_agreement!.find(
      funding => {
        return funding.funder_id && funding.funder_id.value.toLowerCase() === user.account.toLowerCase()
      },
    )
  );
}
