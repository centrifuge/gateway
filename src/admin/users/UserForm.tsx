import React from 'react';
import { Box, Button, FormField, Text, TextInput } from 'grommet';
import { User } from '../../common/models/user';
import * as Yup from 'yup';
import { Formik } from 'formik';
import { PERMISSIONS } from '../../common/constants';
import { MultipleSelect } from '@centrifuge/axis-multiple-select';
import { Schema } from '../../common/models/schema';
import { mapSchemaNames } from '../../common/schema-utils';

type InviteProps = {
  user: User,
  schemas: Schema[],
  onSubmit?: (user) => void;
  onDiscard?: () => void;
}

export default class UserForm extends React.Component<InviteProps> {

  state = { submitted: false };

  onSubmit = (user: User) => {
    this.props.onSubmit && this.props.onSubmit(user);
  };


  render() {

    const userValidation = Yup.object().shape({
      name: Yup.string()
        .max(40, 'Please enter no more than 40 characters')
        .required('This field is required'),
      email: Yup.string()
        .email('Please enter a valid email')
        .required('This field is required'),
      permissions: Yup.array()
        .required('This field is required'),
      schemas: Yup.array()
        .test({
          name: 'test_schemas',
          test: (function(this, value) {
            if (this.parent.permissions.includes(PERMISSIONS.CAN_MANAGE_DOCUMENTS)) {
              return (value && value.length);
            }
            return true;
          }),
          message: 'This field is required',
        }),
    });

    const { user, schemas } = this.props;

    const { submitted } = this.state;

    const permissionOptions = [
      PERMISSIONS.CAN_MANAGE_USERS,
      PERMISSIONS.CAN_MANAGE_SCHEMAS,
      PERMISSIONS.CAN_MANAGE_DOCUMENTS,
      PERMISSIONS.CAN_VIEW_DOCUMENTS,
    ];


    const schemaOptions = schemas.map(i => i.name);

    return (
      <Box margin={{ vertical: 'medium' }}>
        <Formik
          initialValues={user}
          validateOnBlur={submitted}
          validateOnChange={submitted}
          validationSchema={userValidation}
          onSubmit={(values, { setSubmitting }) => {
            if (!values) return;
            this.onSubmit(values);
            setSubmitting(true);
          }}
        >
          {
            ({
               values,
               errors,
               handleChange,
               handleSubmit,
               setFieldValue,
               dirty,
             }) => (
              <form
                onSubmit={event => {
                  this.setState({ submitted: true });
                  handleSubmit(event);
                }}
              >
                <Box gap="medium">
                  <FormField
                    label="Name"
                    error={errors.name}
                  >
                    <TextInput
                      name="name"
                      value={values.name || ''}
                      onChange={handleChange}
                    />
                  </FormField>
                  <FormField
                    label="Email"
                    error={errors!.email}
                  >
                    <TextInput
                      name="email"
                      value={values!.email}
                      onChange={handleChange}
                    />
                  </FormField>

                  <Box margin={{ bottom: 'medium' }}>
                    <FormField
                      label="Permissions"
                      error={errors!.permissions}
                    >
                      <MultipleSelect
                        value={values.permissions}
                        options={permissionOptions}
                        onChange={(selection) => {
                          setFieldValue('permissions', selection);
                          if (!selection.includes(PERMISSIONS.CAN_MANAGE_DOCUMENTS)) {
                            setFieldValue('schemas', []);
                          }
                        }}
                      />

                    </FormField>
                  </Box>

                  {
                    values.permissions.includes(PERMISSIONS.CAN_MANAGE_DOCUMENTS) && <>
                      {schemaOptions && schemaOptions.length > 0 ?
                        <Box margin={{ bottom: 'medium' }}>
                          <FormField
                            label="Document schemas"
                            error={errors!.schemas}
                          >
                            <MultipleSelect
                              labelKey={'name'}
                              valueKey={'name'}
                              value={mapSchemaNames(values.schemas, schemas)}
                              options={schemas}
                              onChange={(selection) => {
                                setFieldValue('schemas', selection.map(s => s.name));
                              }}
                            />

                          </FormField>
                        </Box>
                        :
                        <Text color={'status-warning'}>No schemas in the database. Please add schemas!</Text>
                      }
                    </>
                  }

                  <Box direction="row" justify={'end'} gap={'medium'}>
                    <Button
                      label="Discard"
                      onClick={this.props.onDiscard}
                    />
                    <Button
                      disabled={!dirty}
                      type="submit"
                      primary
                      label={user._id ? 'Update' : 'Create'}
                    />
                  </Box>
                </Box>
              </form>
            )
          }
        </Formik>
      </Box>
    );
  }
}

