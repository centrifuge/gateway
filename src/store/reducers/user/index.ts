import auth from './auth';
import { combineReducers } from 'redux';
import { httpRequestReducer } from '../http-request-reducer';
import {
  userRegisterAction,
  getAllUsersAction,
  userInviteAction,
  updateUserAction,
} from '../../actions/users';
import { User } from '../../../common/models/user';

const register = httpRequestReducer(userRegisterAction);
const list = httpRequestReducer<User[]>(getAllUsersAction);
const invite = httpRequestReducer(userInviteAction);
const update = httpRequestReducer(updateUserAction);
// TODO this needs to be refactored. The store needs a user list for the invite functionality
export default combineReducers({ auth, register, list, invite, update });
