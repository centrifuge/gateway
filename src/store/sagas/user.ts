import { call, fork, put, take, takeEvery } from 'redux-saga/effects';
import { httpClient } from '../../http-client';
import {
  getAllUsersAction,
  updateUserAction,
  userInviteAction,
  userLoginAction,
  userRegisterAction,
} from '../actions/users';
import { User } from '../../common/models/user';
import routes from '../../routes';
import { push } from 'connected-react-router';
import { alertError } from '../actions/notifications';

export function* loginUser(user: User) {
  try {
    const response = yield call(httpClient.user.login, user);
    yield put({
      type: userLoginAction.success,
      payload: response.data,
    });
  } catch (e) {
    yield put({ type: userLoginAction.fail, payload: e });
  }
}

export function* watchLoginPage(action) {
  yield fork(loginUser, action.user);
  yield take(userLoginAction.success);
  yield put(push(routes.index));
}

export function* registerUser(action) {
  try {
    const user = action.user;
    const response = yield call(httpClient.user.register, user);
    yield put({
      type: userRegisterAction.success,
      payload: response.data,
    });
  } catch (e) {
    yield put({ type: userRegisterAction.fail, payload: e });
  }
}

export function* inviteUser(action) {
  try {
    const user = action.user;
    const response = yield call(httpClient.user.invite, user);
    yield put({
      type: userInviteAction.success,
      payload: response.data,
    });
    // reload the users
    yield put({
      type: getAllUsersAction.start,
    });
  } catch (e) {
    yield put({ type: userInviteAction.fail, payload: e });
    yield put(alertError(
      'Failed to invite user',
      e.message,
      { onConfirmAction: { type: userInviteAction.clearError } },
    ));
  }
}

export function* updateUser(action) {
  try {
    const user = action.user;
    const response = yield call(httpClient.user.update, user);
    yield put({
      type: updateUserAction.success,
      payload: response.data,
    });
    // reload the users
    yield put({
      type: getAllUsersAction.start,
    });
  } catch (e) {

    yield put(alertError(
      'Failed to update user',
      e.message,
      { onConfirmAction: { type: updateUserAction.clearError } },
    ));
    yield put({ type: updateUserAction.fail, payload: e });

  }
}

export function* getAllUsers() {
  try {
    const response = yield call(httpClient.user.list);
    yield put({
      type: getAllUsersAction.success,
      payload: response.data,
    });
  } catch (e) {
    yield put({ type: userRegisterAction.fail, payload: e });
  }
}

export default {
  watchGetAllUsers: () => takeEvery(getAllUsersAction.start, getAllUsers),
  watchLoginPage: () => takeEvery(userLoginAction.start, watchLoginPage),
  watchUserInvite: () => takeEvery(userInviteAction.start, inviteUser),
  watchUserRegister: () => takeEvery(userRegisterAction.start, registerUser),
  watchUserUpdate: () => takeEvery(updateUserAction.start, updateUser),
};
