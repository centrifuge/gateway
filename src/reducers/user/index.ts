import auth from './auth';
import { combineReducers } from 'redux';
import { httpRequestReducer } from '../http-request-reducer';
import { userRegisterAction } from '../../actions/users';

export const register = httpRequestReducer(userRegisterAction);

export default combineReducers({ auth, register });
