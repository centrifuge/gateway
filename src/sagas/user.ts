import { call, fork, put, take, takeEvery } from 'redux-saga/effects';
import { httpClient } from '../http-client';
import { userLoginAction } from '../actions/users';
import { User } from '../common/models/dto/user';
import routes from '../routes';
import { push } from 'connected-react-router';

export function* loginUser(user: User) {
  try {
    const response = yield call(httpClient.user.login, user);
    yield put({
      type: userLoginAction.success,
      payload: response.data,
    });
  } catch (e) {
    yield put({ type: userLoginAction.fail, payload: e });
  }
}

export function* watchLoginPage(action) {
  yield fork(loginUser, action.user);
  yield take(userLoginAction.success);
  yield put(push(routes.invoices.index));
}

export default {
  watchLoginPage: () => takeEvery(userLoginAction.start, watchLoginPage),
};
