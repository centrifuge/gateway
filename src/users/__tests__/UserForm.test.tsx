import React from 'react';
import { mount } from 'enzyme';
import UserForm from '../UserForm';
import { User } from '../../common/models/user';
import { PERMISSIONS } from '../../common/constants';

it('Should render the create state', () => {
  const user = new User();
  const schemas = [];

  const form = mount(
    <UserForm user={user} schemas={schemas}/>,
  );

  expect(form.html()).toMatchSnapshot();
});


it('Should render in edit state with all fields including values for schemas input', () => {
  const user = new User();
  user.name = 'Test User';
  user.email = 'test@centrifuge.io';
  user.permissions = [PERMISSIONS.CAN_MANAGE_SCHEMAS, PERMISSIONS.CAN_MANAGE_DOCUMENTS];
  user.schemas = ['option-1'];
  const schemas = [{ name: 'option-1' }, { name:  'option-2'}] ;

  const form = mount(
    <UserForm user={user} schemas={schemas}/>,
  );
  expect(form.html()).toMatchSnapshot();
});
