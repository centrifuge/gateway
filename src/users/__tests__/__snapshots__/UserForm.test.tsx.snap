// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render in edit state with all fields including values for schemas input 1`] = `"<div class=\\"StyledBox-sc-13pk1d4-0 laNWCK\\"><form><div class=\\"StyledBox-sc-13pk1d4-0 IRSNj\\"><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Name</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><div class=\\"StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm\\"><input autocomplete=\\"off\\" name=\\"name\\" class=\\"StyledTextInput-sc-1x30a0s-0 duvNfq\\" value=\\"Test User\\"></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 irlfVP\\"></div><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Email</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><div class=\\"StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm\\"><input autocomplete=\\"off\\" name=\\"email\\" class=\\"StyledTextInput-sc-1x30a0s-0 duvNfq\\" value=\\"test@centrifuge.io\\"></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 irlfVP\\"></div><div class=\\"StyledBox-sc-13pk1d4-0 iFWBSR\\"><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Permissions</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><button class=\\"Select__StyledSelectDropButton-sc-17idtfo-1 cvpccN StyledButton-sc-323bzc-0 jfbKsv\\" aria-label=\\"Open Drop\\" type=\\"button\\"><div class=\\"StyledBox-sc-13pk1d4-0 erazuK\\"><div class=\\"StyledBox-sc-13pk1d4-0 XINUX\\"><div class=\\"StyledBox-sc-13pk1d4-0 ggrNVD\\"><div class=\\"StyledBox-sc-13pk1d4-0 jucMNU\\"><span style=\\"line-height: 1;\\" class=\\"StyledText-sc-1sadyjn-0 fWSbXS\\">can_manage_schemas</span><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 gufrXq\\"></div><div class=\\"StyledBox-sc-13pk1d4-0 IRSNj\\"><svg aria-label=\\"Close\\" viewBox=\\"0 0 24 24\\" class=\\"StyledIcon-ofa7kd-0 fjELgu\\"><path fill=\\"none\\" stroke=\\"#000\\" stroke-width=\\"2\\" d=\\"M3,3 L21,21 M3,21 L21,3\\"></path></svg></div></div><div class=\\"StyledBox-sc-13pk1d4-0 jucMNU\\"><span style=\\"line-height: 1;\\" class=\\"StyledText-sc-1sadyjn-0 fWSbXS\\">can_manage_documents</span><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 gufrXq\\"></div><div class=\\"StyledBox-sc-13pk1d4-0 IRSNj\\"><svg aria-label=\\"Close\\" viewBox=\\"0 0 24 24\\" class=\\"StyledIcon-ofa7kd-0 fjELgu\\"><path fill=\\"none\\" stroke=\\"#000\\" stroke-width=\\"2\\" d=\\"M3,3 L21,21 M3,21 L21,3\\"></path></svg></div></div></div></div><div style=\\"min-width: auto;\\" class=\\"StyledBox-sc-13pk1d4-0 eLQtUu\\"><svg aria-label=\\"FormDown\\" viewBox=\\"0 0 24 24\\" class=\\"StyledIcon-ofa7kd-0 dSllCE\\"><polyline fill=\\"none\\" stroke=\\"#000\\" stroke-width=\\"2\\" points=\\"18 9 12 15 6 9\\"></polyline></svg></div></div></button></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 irlfVP\\"></div><div class=\\"StyledBox-sc-13pk1d4-0 iFWBSR\\"><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Document schemas</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><button class=\\"Select__StyledSelectDropButton-sc-17idtfo-1 cvpccN StyledButton-sc-323bzc-0 jfbKsv\\" aria-label=\\"Open Drop\\" type=\\"button\\"><div class=\\"StyledBox-sc-13pk1d4-0 erazuK\\"><div class=\\"StyledBox-sc-13pk1d4-0 XINUX\\"><div class=\\"StyledBox-sc-13pk1d4-0 ggrNVD\\"><div class=\\"StyledBox-sc-13pk1d4-0 jucMNU\\"><span style=\\"line-height: 1;\\" class=\\"StyledText-sc-1sadyjn-0 fWSbXS\\">option-1</span><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 gufrXq\\"></div><div class=\\"StyledBox-sc-13pk1d4-0 IRSNj\\"><svg aria-label=\\"Close\\" viewBox=\\"0 0 24 24\\" class=\\"StyledIcon-ofa7kd-0 fjELgu\\"><path fill=\\"none\\" stroke=\\"#000\\" stroke-width=\\"2\\" d=\\"M3,3 L21,21 M3,21 L21,3\\"></path></svg></div></div></div></div><div style=\\"min-width: auto;\\" class=\\"StyledBox-sc-13pk1d4-0 eLQtUu\\"><svg aria-label=\\"FormDown\\" viewBox=\\"0 0 24 24\\" class=\\"StyledIcon-ofa7kd-0 dSllCE\\"><polyline fill=\\"none\\" stroke=\\"#000\\" stroke-width=\\"2\\" points=\\"18 9 12 15 6 9\\"></polyline></svg></div></div></button></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 irlfVP\\"></div><div class=\\"StyledBox-sc-13pk1d4-0 fSSGpo\\"><button type=\\"button\\" class=\\"StyledButton-sc-323bzc-0 kHrfdt\\">Discard</button><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 hyyRSo\\"></div><button disabled=\\"\\" type=\\"submit\\" class=\\"StyledButton-sc-323bzc-0 hdVZRh\\">Create</button></div></div></form></div>"`;

exports[`Should render the create state 1`] = `"<div class=\\"StyledBox-sc-13pk1d4-0 laNWCK\\"><form><div class=\\"StyledBox-sc-13pk1d4-0 IRSNj\\"><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Name</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><div class=\\"StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm\\"><input autocomplete=\\"off\\" name=\\"name\\" class=\\"StyledTextInput-sc-1x30a0s-0 duvNfq\\" value=\\"\\"></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 irlfVP\\"></div><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Email</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><div class=\\"StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm\\"><input autocomplete=\\"off\\" name=\\"email\\" class=\\"StyledTextInput-sc-1x30a0s-0 duvNfq\\" value=\\"\\"></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 irlfVP\\"></div><div class=\\"StyledBox-sc-13pk1d4-0 iFWBSR\\"><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Permissions</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><button class=\\"Select__StyledSelectDropButton-sc-17idtfo-1 cvpccN StyledButton-sc-323bzc-0 jfbKsv\\" aria-label=\\"Open Drop\\" type=\\"button\\"><div class=\\"StyledBox-sc-13pk1d4-0 erazuK\\"><div class=\\"StyledBox-sc-13pk1d4-0 XINUX\\"><div class=\\"StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm\\"><input autocomplete=\\"off\\" placeholder=\\"Select\\" multiple=\\"\\" tabindex=\\"-1\\" type=\\"text\\" readonly=\\"\\" class=\\"Select__SelectTextInput-sc-17idtfo-0 lmyIUy StyledTextInput-sc-1x30a0s-0 jfnKzs\\" value=\\"\\"></div></div><div style=\\"min-width: auto;\\" class=\\"StyledBox-sc-13pk1d4-0 eLQtUu\\"><svg aria-label=\\"FormDown\\" viewBox=\\"0 0 24 24\\" class=\\"StyledIcon-ofa7kd-0 dSllCE\\"><polyline fill=\\"none\\" stroke=\\"#000\\" stroke-width=\\"2\\" points=\\"18 9 12 15 6 9\\"></polyline></svg></div></div></button></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 irlfVP\\"></div><div class=\\"StyledBox-sc-13pk1d4-0 fSSGpo\\"><button type=\\"button\\" class=\\"StyledButton-sc-323bzc-0 kHrfdt\\">Discard</button><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 hyyRSo\\"></div><button disabled=\\"\\" type=\\"submit\\" class=\\"StyledButton-sc-323bzc-0 hdVZRh\\">Create</button></div></div></form></div>"`;
